### PR TITLE
MCP: Remove all non-canonical fields impacting RBAC from tools/call request 

### PR DESCRIPTION
**Pull request by kopiczko**
Follow up of https://github.com/gravitational/teleport/pull/66989, prompted by [Codex review on the backport](https://github.com/gravitational/teleport/pull/67174#discussion_r3316973077).

To mitigate a potential RBAC bypass to a vulnerable upstream MCP server by removing all non-lowercase name parameters in tools/call requests. It may happen that a MCP server using case-insensitive Go "json" package decodes "Name" while the RBAC is always performed on performed on lowercase "name" (conforming with the MCP spec) if both are present in the request.

See https://github.com/gravitational/teleport/pull/67282 for a ID fix for similar class of problems.

No changelog because I'm planning to combine those PRs for backport.

## Backports

- https://github.com/gravitational/teleport/pull/68500

### Test Environment

- Upstream MCP server printing incoming requests with 3 different transports: streamable-HTTP, SSE, stdio
- Connection to the upstream MCP server is done with `tsh mcp connect $NAME`

### Test Cases

- [x] For streamable-HTTP transport
   - [x] Verify the server responds to a normal initialization flow and tools/list call
   - [x] Verify the server responds to a tools/call for an RBAC-allowed tool
   - [x] Verify the tools/call request is not forwarded to the upstream MCP server for RBAC-denied tool
   - [x] Verify an extra "Name" parameter is removed from the forwarded request
   - [x] Verify an extra "Method" field is removed from the forwarded request
   - [x] Verify an extra "Params" field is removed from the forwarded request
   - [x] Verify an extra "ID" field is removed from the forwarded request
 - [x] Do the same for SSE transport
 - [x] Do the same for stdio transport

### CODE REVIEW

File: lib/srv/mcp/http.go [hunk 1]
```diff
@@ -219,9 +225,10 @@ func (t *streamableHTTPTransport) handleMCPMessage(r *http.Request) (*http.Respo
 		return nil, trace.BadParameter("invalid request body %v", err)
 	}
 
+	var mcpRequest *mcputils.JSONRPCRequest
 	switch {
 	case baseMessage.IsRequest():
-		mcpRequest := baseMessage.MakeRequest()
+		mcpRequest = baseMessage.MakeRequest()
 		if errResp := t.sessionHandler.processClientRequest(r.Context(), mcpRequest); errResp != nil {
```
Comment (greedy52) [inline lib/srv/mcp/http.go]: nit: a little bit sneaky that the process function changes input.  might be better to explicitly return the request.

File: lib/srv/mcp/session.go [hunk 2]
```diff
@@ -370,3 +377,11 @@ func toError(e mcp.JSONRPCError) error {
 	}
 	return e.Error.AsError()
 }
+
+func sanitizeParamsName(p mcputils.JSONRPCParams) {
```
Comment (greedy52) [inline lib/srv/mcp/session.go]: nit: should this sanitize function be declared in mcputils? could even hook it up via UnmarshalJSON lol

File: lib/srv/mcp/session_test.go [hunk 3]
```diff
@@ -163,6 +164,39 @@ func Test_sessionHandler(t *testing.T) {
 				})
 			}
 
+			for _, allowedTool := range tt.allowedTools {
```
Comment (greedy52) [inline lib/srv/mcp/session_test.go:167]: nit: i don't mind this but might be better to test these against `sanitizeParamsName` directly

**Review by greedy52 — COMMENTED** [review]
LGTM overall. a few nit comments

File: lib/srv/mcp/session_test.go (hunk 3 above)
Comment (kopiczko) [reply lib/srv/mcp/session_test.go:167]: My thinking was it's more important what we send to upstream rather than testing particular functions. How we get there is not that important. Testing `sanitizeParamsName` alone doesn't guarantee it's going to be called anywhere and the function alone is trivial.

File: lib/srv/mcp/session.go (hunk 2 above)
Comment (kopiczko) [reply lib/srv/mcp/session.go]: I think it's matter of taste, but I like it being called explicitly rather than hiding inside UnmarshalJSON.

As for the package - I don't know. It isn't going to be a big job if it should be moved.

File: lib/srv/mcp/http.go (hunk 1 above)
Comment (kopiczko) [reply lib/srv/mcp/http.go]: Addressed in [Change processRequest signature to return mutated request](https://github.com/gravitational/teleport/pull/67262/commits/daced9fce64a87c6b1d28369481563737a9da75d) to conform to `append` signature.

**Comment by kopiczko** [comment]
@greedy52 PTALA

File: lib/srv/mcp/http.go [hunk 4]
```diff
@@ -219,22 +225,26 @@ func (t *streamableHTTPTransport) handleMCPMessage(r *http.Request) (*http.Respo
...
 	switch {
 	case baseMessage.IsRequest():
 		mcpRequest := baseMessage.MakeRequest()
-		if errResp := t.sessionHandler.processClientRequest(r.Context(), mcpRequest); errResp != nil {
+		mcpRequest, errResp := t.sessionHandler.processClientRequest(r.Context(), mcpRequest)
+		if errResp != nil {
 			t.emitRequestEvent(t.parentCtx, mcpRequest, eventWithError(toError(*errResp)), eventWithHeader(r.Header))
 			return t.handleRequestError(r, *errResp)
 		}
+		resp, sendReqErr = t.rewriteAndSendRequestWithBody(r, mcpRequest)
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/srv/mcp/http.go]: That's a valid comment. I think we can reduce it to rewriting the request body only when the message needs to be modified.

[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/mcp/http.go]: or maybe just reject tools/call with `nAME` because it looks suspicious.

[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/mcp/http.go]: or you can use other lib to work on the json bytes directly like https://github.com/gravitational/teleport/blob/211d2fe196b9fd2666ffee286507f6f1a04e403f/lib/client/mcp/config/config.go#L133

[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/srv/mcp/http.go]: I have some ideas. Changes incoming, but they will be quite substantial unfortunately.

[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/srv/mcp/http.go]: I pushed those two:

- [Mutate request only when necessary + add "method" and "params"](https://github.com/gravitational/teleport/pull/67262/commits/3065d1de742459b639be9779426d1fa4785a6552)
- [Sanitize more fields which can lead to a confused upstream](https://github.com/gravitational/teleport/pull/67262/commits/2aab3af6d99d01ecd385d658a084274b0187a227)

but I think the changes are so large that it makes sense to re-review the whole thing.

**Comment by kopiczko** [comment]
@greedy52 @gabrielcorado please have a look

**Review by gabrielcorado — APPROVED** [review]

File: lib/srv/mcp/sanitize.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+//
+//  4. Trying to confuse upstream MCP server with, non-canonical "id" filed. Without this
+//     sanitization Teleport correctly interprets the message as a notification (messages without
+//     the "id" field are Requests according to the spec), but the upstream server may incorrectly
+//     decode it as a tools/call request and execute the denied_tool.
+//
+//     {"jsonrpc":"2.0","ID":1,"method":"tools/call","params":{"name":"denied_tool"}}
+//
+// It is implemented in a way that prevents corruptions caused by rounding float-precision number
+// or overflowing very large numbers.
```
Comment (greedy52) [inline lib/srv/mcp/sanitize.go:74]: The implementation of `sanitizeRawMCPRequest` is really cool. though it's a little hard to scale (e.g i don't see sanitize or corruption being handled for stdio/sse flows) and maintain long term (as it targets specific fields). 

Maybe we should carefully evaluate this "corruption". Is it a real issue? if yes, should it be fixed in other places? 

For request ID, the special `RequestID` type should be handling things properly. Even if not handled properly, existing logic in stdio/sse would "corrupt" the request id in the same fashion consistently at least 😅 .

One part i think that may cause corruption is `type JSONRPCParams map[string]any`. Maybe that should be redesigned?

WDYT?

Comment (kopiczko) [reply lib/srv/mcp/sanitize.go:74]: You're correct here that it doesn't span to the other transports. It only applies to the HTTP transport where we forward raw requests.

If you zoom out the problem is even worse, because we actually decode raw requests we forwarded for audit events, so se may with a different request in the audit logs than we actually forwarded.

So I completely see your point, but my goal with this PR is to fix the issue with non-canonical names and not introduce any regression (in form of potentially corrupted floats and big numbers).

But I agree it'd be good to make a decision and unify this approach across all transport. As you said a good example is `type JSONRPCParams map[string]any` which is inconsistently send across transports and may be corrupted. I think it's a separate PR though. Would you agree?

Comment (greedy52) [reply lib/srv/mcp/sanitize.go:74]: i thought stdio and sse is also vulnerable to `nAme`

Comment (kopiczko) [reply lib/srv/mcp/sanitize.go:74]: I think both stdio and sse call `sessionHandler.onClientRequest` which calls `sanitizeMCPRequestParams` (defined above), before passing the request to `serverRequestWriter.WriteMessage`. They are not affected by the top level fields because JSONRPCRequest.UnmarshalJSON is case-sensitive. I tried to explain that in the `sanitizeMCPRequestParams` godoc.

I extended `Test_sessionHandler` in `session_test.go` to cover that. Do you think that requires also extending the integration test?

Comment (greedy52) [reply lib/srv/mcp/sanitize.go:74]: thanks. i missed that part in session handler. if that is case, i would prefer streamable-http to get the same treatment like always marshalling `mcputils.JSONRPCRequest` to keep all transports consistent when possible. i can also do the follow up on fixing `JSONRPCParams` right away so we can backport them together. 

as for audit event, it supposes to capture what client sends, not what server will receive. so i would say closer to the original content, the better it is. imagine audit event captures `nAme`.

File: lib/srv/mcp/helpers_test.go [hunk 6]
```diff
@@ -494,15 +477,128 @@ func newStreamableMCPServer(t *testing.T, upstream *mcpserver.MCPServer, role ty
...
+	requests []*http.Request
+}
+
+func (r *testHTTPRequestRecorder) registerHttpRequest(t *testing.T, req *http.Request) {
+	t.Helper()
+
+	data, err := utils.GetAndReplaceRequestBody(req)
 	require.NoError(t, err)
+	req = req.Clone(t.Context())
+	utils.WriteRequestBody(req, data)
```
Comment (greedy52) [inline lib/srv/mcp/helpers_test.go]: is this `utils.WriteRequestBody` necessary?

File: lib/utils/http.go [hunk 7]
```diff
@@ -83,6 +81,14 @@ func ReplaceRequestBody(req *http.Request, newBody io.ReadCloser) error {
 	return nil
 }
 
+// WriteRequestBody (over)writes the new data into the given request's body. It doesn't check if
+// the previous req.Body exists or if it's closed. It's the callers responsibility to cleanup the
+// previous body.
+func WriteRequestBody(req *http.Request, data []byte) {
```
Comment (greedy52) [inline lib/utils/http.go]: > It's the callers responsibility to cleanup the previous body.

nit: I feel i will misuse this function as likely i won't be reading the full details. should we just use the existing `ReplaceRequestBody` or make the function name more obvious? or maybe a helper that does get and rewrite in the same func by taking a callback?

File: lib/srv/mcp/helpers_test.go (hunk 6 above)
Comment (kopiczko) [reply lib/srv/mcp/helpers_test.go]: From `Clone` godoc: `// Clone only makes a shallow copy of the Body field.`

File: lib/srv/mcp/sanitize.go (hunk 5 above)
Comment (greedy52) [reply lib/srv/mcp/sanitize.go:74]: rough draft https://github.com/gravitational/teleport/pull/67685

File: lib/utils/http.go (hunk 7 above)
Comment (kopiczko) [reply lib/utils/http.go]: Addressed in [Drain the request body just in case before overwriting](https://github.com/gravitational/teleport/pull/67262/commits/7b8bb4ccfdbd8c08a235b46b7d63eefed76ec66b)

File: lib/srv/mcp/sanitize.go (hunk 5 above)
Comment (kopiczko) [reply lib/srv/mcp/sanitize.go:74]: I'd even say it we could include the raw request and the forwarded request in the audit event.

File: lib/srv/mcp/sanitize.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+	// Use json.RawMessage, to not mess with the content, e.g. introduce floating-point
+	// corruption caused by rounding.
+	var m map[string]json.RawMessage
+	if err := json.Unmarshal(request, &m); err != nil {
+		return nil, trace.Wrap(err, "unmarshaling MCP Request into a map")
+	}
+
+	// Delete non-canonical "id", "method", "params" top-level message fields.
+	modifiedM := deleteNonCanonicalKey(m, mcputils.IDKey, mcputils.MethodKey, mcputils.ParamsKey)
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/srv/mcp/sanitize.go:90]: We won't be protecting from classifying Notfication messages as Request messages. That'd be too much.

**Comment by kopiczko** [comment]
@greedy52 PTALA

**Comment by greedy52** [comment]
> @greedy52 PTALA

approved. though i think this is likely only a temporary measurement. long term hopefully this lib will get some good refactoring (especially whilting migrating to official mcp lib) so marshalling and audit events are handled more consistently. well, the other way is to deprecate stdio and sse transport, 😉 

please include some manual testing though, at the minimum a regular streamable-http mcp does not have regression issues.

**Review by greedy52 — APPROVED** [review]
